daemon/src/close_cfds.rs

```diff
@@ -0,0 +1,417 @@
+                Ok(None) => {
+                    break;
+                }
+                Err(e) => {
+                    tracing::warn!("Failed to load CFD from database: {e:#}");
+                    continue;
+                }
+            };
+
+            if !cfd.is_closed() {
+                continue;
+            }
+
+            self.db
+                .insert_closed_cfd(cfd)
+                .await
+                .context("Failed to insert closed CFD to database")?
+
+            // TODO: Save event log to a separate table
+
+            // TODO: Delete closed CFD from cfds table
+
+            // TODO: Delete closed CFD events from events table
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = ();
+
+    async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
+        let this = ctx.address().expect("we are alive");
+        self.tasks
+            .add(this.send_interval(CLOSE_CFDS_INTERVAL, || CloseCfds));
+    }
+
+    async fn stopped(self) -> Self::Stop {}
+}
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle(&mut self, _: CloseCfds) {}
+}
+
+struct CloseCfds;
+
+#[derive(Debug, Clone)]
+pub struct Cfd {
+    pub id: OrderId,
+    pub position: Position,
+    pub initial_price: Price,
+    pub taker_leverage: Leverage,
+    pub n_contracts: Contracts,
+    pub counterparty_network_identity: Identity,
+    pub role: Role,
+    pub opening_fee: OpeningFee,
+    pub initial_funding_rate: FundingRate,
+    pub initial_tx_fee_rate: TxFeeRate,
+    pub fee_account: FeeAccount,
+    pub own_script_pubkey: Option<Script>,
+    pub lock: Option<Lock>,
+    pub commit: Option<Txid>,
+    pub collaborative_settlement: Option<CollaborativeSettlement>,
+    pub non_collaborative_settlement: Option<NonCollaborativeSettlement>,
+    pub refund: Option<Refund>,
+    pub close_status: CloseStatus,
+    pub version: u32,
+}
+
+#[derive(Debug, Clone)]
+pub struct Lock {
+    pub txid: Txid,
+    pub dlc_vout: Vout,
+}
+
+#[derive(Debug, Clone)]
+pub struct CollaborativeSettlement {
+    pub txid: Txid,
+    pub vout: Vout,
+    pub payout: Payout,
+    pub price: Price,
+}
+
+#[derive(Debug, Clone)]
+pub struct NonCollaborativeSettlement {
+    pub cet_txid: Txid,
+    pub cet_vout: Vout,
+    pub cet_payout: Payout,
+    pub price: Price,
+}
+
+#[derive(Debug, Clone)]
+pub struct Refund {
+    pub txid: Txid,
+    pub vout: Vout,
+    pub payout: Payout,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum CloseStatus {
+    NotClosed,
+    Collaborative,
+    NonCollaborative,
+    Refund,
+}
+
+impl CloseStatus {
+    fn set(&mut self, new_status: CloseStatus) {
+        if self == &new_status {
+            return;
+        }
+
+        use CloseStatus::*;
+        if let Collaborative | NonCollaborative | Refund = new_status {
+            tracing::warn!("Close transaction status changed from {self} to {new_status}");
+        }
+
+        *self = new_status
+    }
```

Comment:
Can we not imply that through the transactions that are set in the aggregate?

Reply:
That's what I assumed at first, but I thought we could run into trouble. For example, what would happen if (somehow) the CFD had a collaborative settlement and commit transaction? I wanted to avoid having to write any logic around odd scenarios like that one.

I think it's safer to assume that only one of the final transactions will get 3 confirmations. So setting the `CloseStatus` based solely on those events should be much safer.

Reply:
Writing that comment does highlight to me that the warning above should be a `debug_assert` or an `expect` though.